osgar/drivers/tankmaria.py

```diff
@@ -0,0 +1,167 @@
+     "<PWM left> <PWM right> <desired LED status>\n"
+  The value for PWM can in in -255..255.
+  LED status: 0=OFF, 1=green, 2=red.
+
+  The board regularly responds (20Hz?) with status line:
+      "<32bit unsigned int timestamp in milliseconds> <encoder left> <encoder right> <buttons mask>\n"
+  The encoders values are 32bit signed int and accumulative.
+  The buttons mask is currently not used (first remote tests are with Emergency STOP pressed).
+
+  The plan is to abandon this text protocol and use binary structures inspired by ATmega8board.
+"""
+
+import math
+
+from osgar.node import Node
+from osgar.bus import BusShutdownException
+
+
+ENC_SCALE = 0.00052
+WHEEL_DISTANCE = 0.4  # 50cm long, 40cm wide
+
+GREEN_LED = 1
+RED_LED = 2
+
+
+class MariaTimer(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register('tick')
+        self.sleep_time = config['sleep']
+
+    def run(self):
+        try:
+            while self.is_alive():
+                self.publish('tick', None)
+                self.sleep(self.sleep_time)
+
+        except BusShutdownException:
+            pass
+
+
+class RobotTankMaria(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register('pose2d', 'emergency_stop', 'encoders', 'raw')
+
+        # commands
+        self.desired_speed = 0.0  # m/s
+        self.desired_angular_speed = 0.0
+
+        # status
+        self.emergency_stop = None  # uknown state
+        self.pose = (0.0, 0.0, 0.0)  # x, y in meters, heading in radians (not corrected to 2PI)
+        self.buttons = None
+
+        self.verbose = False  # should be in Node
+        self.buf = b''
+        self.encoder = None
+        self.led = GREEN_LED
+        self.milisec = None
+
+    def send_pose(self):
+        x, y, heading = self.pose
+        self.publish('pose2d', [round(x*1000), round(y*1000),
+                                round(math.degrees(heading)*100)])
+
+
+    def compute_pose(self, left, right):
+        """Update internal pose with 'dt' step"""
+        if left is None or right is None:
+            return False, None, None
+        x, y, heading = self.pose
+
+        metricL = ENC_SCALE * left  # dt is already part of ENC_SCALE
+        metricR = ENC_SCALE * right
+
+        dist = (metricL + metricR)/2.0
+        angle = (metricR - metricL)/WHEEL_DISTANCE
+
+        # advance robot by given distance and angle
+        if abs(angle) < 0.0000001:  # EPS
+            # Straight movement - a special case
+            x += dist * math.cos(heading)
+            y += dist * math.sin(heading)
+            #Not needed: heading += angle
+        else:
+            # Arc
+            r = dist / angle
+            x += -r * math.sin(heading) + r * math.sin(heading + angle)
+            y += +r * math.cos(heading) - r * math.cos(heading + angle)
+            heading += angle # not normalized
```

Comment:
Is there a reason why we do not normalize the heading?

Reply:
In general I do not normalize heading for `Pose2D` from odometry. This way it contains information that the robot turned twice 360 degrees for example, which is information not available from compass for example. So this is intentional, but feel free to argue.

Reply:
Just append this explanation into the code and we are good to go :+1: